add claimtypes.email provider to auth sample & article (#10691)
aspnet/Docs: aspnetcore/signalr/authn-and-authz.md

Add a new class that implements `IUserIdProvider` and retrieve one of the claims from the user to use as the identifier. For example, to use the "Name" claim (which is the Windows username in the form `[Domain]\[Username]`), create the following class:

> [!code-csharp[Identity.Name-based provider](authn-and-authz/sample/nameuseridprovider.cs?range=5-11)]

Reviewer: Prefer "?name=region-name-here" over "?range=". This requires you to wrap the code in a region in the sample code. It makes things more maintainable moving forward. Line ranges are often not updated as the sample evolves.

Windows Authentication is only supported by the browser client when using Microsoft Internet Explorer or Microsoft Edge.

> ### Using Claims to Customize Identity Handling

Reviewer: ```suggestion
### Use Claims to customize Identity handling
```

> If you're using organizational or social authentication in your apps and you'd prefer to have more fine-grained control over how SignalR identifies users, you can implement `IUserIdProvider` to return values from one of the claims collected by the identity provider. 

Author: Makes sense! Will augment.

> The sample code demonstrates how you would use claims to select the user's email address as the identifying property. 

Reviewer 2: Sentence 1 "The sample code"
Sentence 2 "In the sample code"
Sentence 3 "The sample code"

The only suggestion I have is to only keep "sample code" for the first sentence and the other sentences should be fine without it?

Author: egads! agreed

> The code below demonstrates how you would use Claims to select the user's email address as the identifying property. 

Reviewer 3: The following code 
("below" is frowned upon because it doesn't make sense when read by screen readers)

demonstrates how ~~you would use Claims ~~ to select 

(if you keep the ref to claims here:  Claims --> claims)

Author: good feedback, thanks for that. i wasn't aware of those nuances but thanks for bringing them to my attention.

> [!code-csharp[Email provider](authn-and-authz/sample/EmailBasedUserIdProvider.cs?name=EmailBasedUserIdProvider)]

Reviewer 2: We used to have a note saying that the identifier should be unique

> [!code-csharp[Email claim-based provider](authn-and-authz/sample/EmailBasedUserIdProvider.cs?range=6-12)]

Reviewer: See earlier feedback regarding regions instead of ranges.

> In the sample code, the account registration code contains the following line of code, which adds a claim with type `ClaimsTypes.Email` to the ASP.NET identity database. 

Reviewer 2: "following line**s** of code"

> In the sample code, the account registration code contains the following line of code, which adds a claim with type `ClaimsTypes.Email` to the ASP.NET identity database. 

Reviewer: ```suggestion
The sample app's account registration code contains the following, which adds a claim with type `ClaimsTypes.Email` to the ASP.NET Core Identity database:
```

> [!code-csharp[Adding the email to the ASP.NET identity claims](authn-and-authz/sample/pages/account/Register.cshtml.cs?range=64)]

Reviewer: See earlier feedback.

> The sample code contains both the name and email implementations. Register this component in your `Startup.ConfigureServices` method **after** the call to `.AddSignalR`. 

Reviewer 2: Doesn't need to be after AddSignalR anymore I believe.

Reviewer: @BrennanConroy Was that a change in 2.2?

Reviewer 2: Nope, in a patch in 2.1

> The sample code contains both the `Identity.Name`-specific and email-specific implementations. Register this component in your `Startup.ConfigureServices` method **after** the call to `.AddSignalR`. 

Reviewer 3: `Identity.Name`-specific and email-specific implementations. --> name and email implementations.

Author: done!

Windows Authentication is only supported by the browser client when using Microsoft Internet Explorer or Microsoft Edge.

### Using Claims to Customize Identity Handling

If you're using organizational or social authentication in your apps and you'd prefer to have more fine-grained control over how SignalR identifies users, you can implement `IUserIdProvider` to return values from one of the claims collected by the identity provider. 

The code below demonstrates how you would use Claims to select the user's email address as the identifying property. 

```csharp

> public class EmailBasedUserIdProvider : IUserIdProvider

Reviewer 2: Reference the code from the sample

{
    public virtual string GetUserId(HubConnectionContext connection)
    {
        return connection.User?.FindFirst(ClaimTypes.Email)?.Value;
    }
}
```

In the sample code, the account registration code contains the following line of code, which adds a claim with type `ClaimsTypes.Email` to the ASP.NET identity database. 

```csharp

> await _userManager.AddClaimAsync(user, new Claim(ClaimTypes.Email, Input.Email));

Reviewer 2: Same

Author: will do for all of them now
